### PR TITLE
refactoring energy corrections outside qpmin:qpmax for RPA

### PR DESCRIPTION
**Pull request by baumeier**
Implements #557

### CODE REVIEW

**Comment by baumeier** [comment]
@votca-bot format

**Comment by baumeier** [comment]
@votca-bot changelog: refactoring energy corrections for RPA

**Review by JensWehner — APPROVED** [review]
Thanks

File: src/libxtp/gwbse/rpa.cc [hunk 1]
```diff
@@ -33,28 +33,45 @@ void RPA::UpdateRPAInputEnergies(const Eigen::VectorXd& dftenergies,
...
+  _energies.segment(0, qpmin - _rpamin).array() -= max_correction_occ;
+  _energies.segment(offset_qp_virt, _rpamax - qpmax).array() +=
       max_correction_virt;
 }
 
+double RPA::getMaxCorrection(const Eigen::VectorXd& dftenergies, Index min,
+                             Index max) {
+
+  Index range = max - min + 1;
+  Eigen::VectorXd corrections = _energies.segment(min - _rpamin, range) -
```
Comment (JensWehner) [inline src/libxtp/gwbse/rpa.cc:69]: `_energies` has a different size then `dftenergies` 
Please check this again.

Comment (JensWehner) [reply src/libxtp/gwbse/rpa.cc:69]: _energies should be rpatotal whereas dftenergies is mo.size() large

**Review by JensWehner — CHANGES_REQUESTED** [review]
It needs a test too, which should be easy to verfiy using `const Eigen::VectorXd& getRPAInputEnergies() const { return _energies; }`

**Comment by junghans** [comment]
@votca-bot copyright

**Comment by baumeier** [comment]
ping @JensWehner :)

File: include/votca/xtp/rpa.h [hunk 2]
```diff
@@ -92,6 +93,12 @@ class RPA {
   Eigen::MatrixXd Calculate_H2p_ApB() const;
   Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> Diagonalize_H2p_C(
       const Eigen::MatrixXd& C) const;
+
+  void ShiftUncorrectedEnergies(const Eigen::VectorXd& dftenergies,
+                                 Index qpmin, Index gwsize);
+
+  double getMaxCorrection(const Eigen::VectorXd& dftenergies,
+                                 Index min, Index max);
```
Comment (JensWehner) [inline include/votca/xtp/rpa.h:101]: should b
```suggestion
                                 Index min, Index max)const;
```

File: src/libxtp/gwbse/rpa.cc [hunk 3]
```diff
@@ -33,28 +33,44 @@ void RPA::UpdateRPAInputEnergies(const Eigen::VectorXd& dftenergies,
...
+  double max_correction_virt = getMaxCorrection(dftenergies, lumo, qpmax);
+
+  // shift energies
+  _energies.head(qpmin).array() -= max_correction_occ;
+  _energies.tail(_rpamax - qpmax).array()+=
       max_correction_virt;
 }
 
+double RPA::getMaxCorrection(const Eigen::VectorXd& dftenergies, Index min,
+                             Index max) {
```
Comment (JensWehner) [inline src/libxtp/gwbse/rpa.cc:65]:
```suggestion
                             Index max) const{
```

**Review by JensWehner — CHANGES_REQUESTED** [review]
two const.

**Comment by baumeier** [comment]
@votca-bot format

**Review by JensWehner — APPROVED** [review]
I looked at it for 10 min and now I am more confused than before with all the indeces, but I think it is correct.

**Pull request merged**